enhancements/20220128-single-node-support-bundle.md

```diff
@@ -0,0 +1,167 @@
+
+No API changes.
+
+## Design
+
+### Implementation Overview
+
+We leverage the `supportconfig` command in [openSUSE supportutils](https://github.com/openSUSE/supportutils) package.
+
+The `supportconfig` command is the standard utility to collect system information for SUSE distributions. It already contains much useful information and some basic health checks, so we don't need to reinvent wheels. The following text contains the file list in a `supportconfig` tarball, we can see there are many components:
+
+```shell
+node1:/tmp/scc_node1_220128_0552 # ls
+basic-environment.txt	open-files.txt
+basic-health-check.txt	pam.txt
+boot.txt		print.txt
+cimom.txt		proc.txt
+crash.txt		rpm.txt
+cron.txt		rpm-verify.txt
+dhcp.txt		samba.txt
+dns.txt			sar.txt
+docker			security-apparmor.txt
+docker.txt		security-audit.txt
+env.txt			shell_history.txt
+etc.txt			slert.txt
+fs-autofs.txt		slp.txt
+fs-btrfs.txt		smt.txt
+fs-diskio.txt		ssh.txt
+fs-files.txt		sssd.txt
+fs-iscsi.txt		summary.xml
+fs-smartmon.txt		supportconfig.txt
+fs-softraid.txt		sysconfig.txt
+hardware.txt		sysfs.txt
+ldap.txt		systemd.txt
+lvm.txt			transactional-update.txt
+memory.txt		tuned.txt
+messages.txt		udev.txt
+modules.txt		updates.txt
+mpio.txt		web.txt
+nfs.txt			y2log.txt
+ocfs2.txt
+```
+
+The components can be included or excluded according to our needs.
+
+The `supportconfig` command can be extended via plugins. We add new plugins to collect various component information:
+ - `harvester`: collect installer logs, cloud-init files, and other useful information that is not included in standard supportconfig modules.
+- `rke2`: collect RKE2 configs and logs.
+- `rancherd`: collect rancherd configs, plans, and logs.
+- `rancher-system-agent`: collect `rancher-system-agent` configs, plans, and logs.
+
+Take `harvester` plugin as the example: a generated node support bundle contains a file `plugin-harvester.txt` that contains information:
+
+```
+#==[ Plugin ]=======================================#
+# Output: /usr/lib/supportconfig/plugins/harvester
+
+#==[ Command ]======================================#
+# /usr/lib/supportconfig/plugins/harvester
+
+#==[ Section Header ]===============================#
+# Harvester: installed mode
+
+
+#==[ Section Header ]===============================#
+# Luet packages
+
+#==[ Command ]======================================#
+# /usr/bin/luet search --installed
+> toolchain/luet-cosign-0.0.10-3
+ -> Category: toolchain
+ -> Name: luet-cosign
+ -> Version: 0.0.10-3
+ -> Description:
+ -> Repository: system
+ -> Uri:
+ -> Installed: true
+```
+
+The plugin creates a `harvester` directory to store configs and logs:
+
+```
+# find harvester/
+harvester/
+harvester/oem
+harvester/oem/99_custom.yaml
+harvester/oem/harvester.config
+harvester/console.log
+```
+
+Similar mechanisms apply to other plugins as well.
+
+Note the config files might contain sensitive information like passwords. We need to remove them when generating a support bundle.
+
```

Comment:
# Upload support config
Introduce support for NFS / Object store based upload options for support config.
This will make it easier to collate support configs at a common store, and may be useful for MaaS capability.